prevent accidental editing of metadata from track table

Be-ing opened the pull request:
Before, it was easy to accidentally edit metadata, which is especially dangerous now that Mixxx supports writing metadata back to file tags.

Unfortunately the default F2 keyboard shortcut interferes with this. It works if keyboard shortcuts are turned off. I am not sure what to do about that or if anything should be done. The default keyboard shortcut for F2 is rate_perm_up, which makes such big jumps in the tempo that it's not really useful. I am hesitant to make a change to the default keyboard mapping this close to the 2.1 release. We may leave this known issue unresolved for 2.1.

Fixing https://bugs.launchpad.net/mixxx/+bug/1743238

Be-ing committed:
only allow editing of metadata from track table after pressing F2

Before, it was easy to accidentally edit metadata, which is
especially dangerous now that Mixxx supports writing metadata
back to file tags.

ronso0 commented:
I use this quick edit functionality a lot, mostly for making title/artist more compact, but also for editing HotCue hints in track comments which show up in decks.
For me, Fn+F2+click makes this simple action too complex. Also my Fn button behaves weird in xfce4 when accidentially hitting it twice, and I'm glad I didn't have to use Fn combos in Mixxx so far..

So I do support Ron's request for a respective option, probably placed in Settings > Library > Miscellaneous.

Be-ing commented:
It seems that the issue is caused by double clicking being used both for editing items in the table and for loading tracks to decks or AutoDJ. I presume if you want double clicking to start editing track metadata, you do not want to load it to a deck or the AutoDJ queue. Am I understanding that correctly, @ronso0? If so, we can rename the "Track Load Action" option to "Track Double Click Action" and add a 4th option for editing metadata. Regardless of which option is chosen, pressing F2 should work for editing metadata.

ronso0 commented:
Yes, double clicking is the issue. Can we increase the click delay which distinguishes two separate clicks from a regular double click? So it's Click1 > 1sec > Click 2 = edit metadata

I'll test this branch and come back.

Be-ing commented:
Okay, I implemented that option. Regardless of which double click action is configured, pressing F2 allows editing of the metadata.

ronso0 commented:
> Okay, I implemented that option. Regardless of which double click action is configured, pressing F2 allows editing of the metadata.

Thanks, works fine.

ronso0 commented:
> Maybe with some ugly hacks, but I don't think it's appropriate for an application to be responsible for this. Look for an option in the settings of your desktop environment to change that.

It was not about changing that behaviour globally but to make it easier for users to have both double click > load and two delayed clicks for editing. I supposed Mixxx was dealing with those delays, not the OS

foss- commented:
just wanted to mention: on macOS alt single-click / alt tap (trackpad) goes into editing a files name in finder (file browser).

Looking forward to this, since the accidental edits in mixxx library cause mixxx to basically become unusable due to broken wavefrom display on macOS: [Lauchpad 1665583](https://bugs.launchpad.net/mixxx/+bug/1665583).

foss- commented:
on macOS alt single-click / alt tap (trackpad) goes into editing a files name in finder (file browser). Looking forward to this, since accidental edits in mixxx library cause mixxx to basically become unusable due to broken wavefrom display on macOS: [Lauchpad 1665583](https://bugs.launchpad.net/mixxx/+bug/1665583).

Be-ing commented:
Thank you for testing @ronso0. I think we have come up with a good solution here.

@foss-, I am not certain how this works on macOS. This is using Qt's "EditKey" abstraction. I presume this uses Alt + click on macOS if that is what is normal on that OS, but I am not sure and there does not seem to be any documentation specifically about that.

Be-ing commented:
> It was not about changing that behaviour globally but to make it easier for users to have both double click > load and two delayed clicks for editing. I supposed Mixxx was dealing with those delays, not the OS

I am doubtful there would be a way to do that which would not be prone to accidents.

daschuer commented on src/preferences/dialog/dlgpreflibrarydlg.ui:
```diff
@@ -258,16 +258,9 @@
      </property>
      <property name="title">
-      <string extracomment="Sets default action when double-clicking a track in the library.">Track Load Action</string>
+      <string extracomment="Sets default action when double-clicking a track in the library.">Track Double Click Action</string>
```
This breaks string freeze

Be-ing replied:
Yes, it does unfortunately. The old text is not really accurate anymore. If you insist I can revert this line and only add the new "Edit metadata" string, but the option description won't entirely make sense.

daschuer commented:
We cannot do it like that because of string freeze. 

How about setting QAbstractItemView::SelectedClicked via a new preferences option? 

Where is the code that enters the edit mode before this PR?

Be-ing commented:
> How about setting QAbstractItemView::SelectedClicked via a new preferences option?

I just tested that. I find it unintuitive and I am afraid it still makes it too easy to accidentally edit metadata.

> Where is the code that enters the edit mode before this PR?

There was none. It was done in the default way by Qt, which is [QAbstractItemView::DoubleClicked|QAbstractItemView::EditKeyPressed](https://github.com/qt/qt/blob/0a2f2382541424726168804be2c90b91381608c6/src/gui/itemviews/qabstractitemview.cpp#L84). This created an odd situation because we had a slot that handled double click events manually, so both our double click action and the editing were triggered.

daschuer commented:
Yes, there is no sense in improveing strings, because it only fixes it for one language and makes it unreadable for all other languages.

Be-ing committed:
add preference option for double clicking track table to edit metadata

Note that the "Track Load Action" header for the preference option in
DlgPrefLibrary does not really make sense anymore, but it is too late to
change that string for 2.1.

Be-ing commented:
Okay, I force pushed a new commit reverting that string.

daschuer commented:
Thank you. LGTM

esbrandt commented:
@daschuer 
Help me out here...

You allowed to break string freeze with the addition of a new string.
But not to change the containing groupbox´ label and tooltip, which are now wrong because of the added string?

Be-ing commented:
@esbrandt, what is the keyboard shortcut for editing the metadata info in macOS?

esbrandt commented:
@Be-ing 
There is a specific key?
Never used one for inline editing, always single clicked on the track.

This lead me to request a similar feature:
[Add shortcut to open Properties editor when a track is selected in library](https://bugs.launchpad.net/mixxx/+bug/1393357)

AFAIK there is no equivalent to the Win/Linux ``F2`` shortcut on macOS.
``Rename`` has no associated shortcut by default, it works by just like in Mixxx - by single clicking on the file.

Be-ing commented:
@foss- suggested alt + click on macOS might work like F2 on Windows and Linux. Can you test that?

esbrandt commented:
Tested macOS 10.13.3

In Finder (macOS file manager)
* ``Alt+Click`` does nothing
* Select file and single click on name enters inline editing mode - alternatively ``cmd+i`` for file info incl. possibility to change the file name.

foss- commented:
Ups. esbrandt is correct. My bad.

Be-ing commented:
Maybe F2 works with Mixxx anyway on macOS? Can you try disabling keyboard shortcuts from the Options menu and pressing F2 with a track selected in the track table?

foss- commented:
Disabled keyboard shotcuts. F2 does nothing when pressed with a track selected in track table.

esbrandt commented:
Even if we somehow force F2 far mac users, it will not work until the user changes their keyboard layout. By default, the function keys access the special function printed on the screens ( e.g. volume up/down, lighten/darken screen etc.)

Note :
With the Touch Bar on some MacBook Pro´s from 2016+ onwards, they replaced the function keys on the keyboard . It changes automatically based on what you’re doing and apps that you’re using. 
For access to function keys F1-F12, you need to hold hold down the Function (fn) key at the keyboard. Touch Bar changes to show the function keys, and returns to its previous state on release (fn).
(untested, no newer macbook available)

foss- commented:
Correct. Fn+F2 is what I tested on an old non-touchbar macbook pro.

ywwg commented:
The change feature was confusing to me and I had trouble figuring out the correct new workflow (I didn't know about F2).  I'm ok with the new behavior, but it would be nice if [some modifier] + click triggered edit mode, maybe Shift?  Or, adding an entry to the right-click menu to edit the current field would be nice.

Be-ing commented:
I think Shift + click or Alt + click could work well. I don't think adding an entry to the right click menu would be very convenient to use because it would require the extra step of looking through the menu and clicking the new entry.

Be-ing commented:
@ywwg, would you like to implement that? I think connecting to the `clicked` signal of WTrackTableView and checking QApplication::keyboardModifiers() should work.

daschuer commented:
Since the selected click is standard in Mac and windows, we should IMHO instead consider to "repair" that. This could be at least an option. 
Do we have an idea why it works in Windows Explorer but not in Mixxx? 
Or is it just am issue of the Linux users who just do not expect it?

Be-ing commented:
What do you mean by "selected click"? Clicking again on the track that is already focused? That would not work because edit mode would be triggered before double clicking on a track.

ywwg commented:
> Since the selected click is standard in Mac and windows, we should IMHO instead consider to "repair" that. This could be at least an option.
> Do we have an idea why it works in Windows Explorer but not in Mixxx?
> Or is it just am issue of the Linux users who just do not expect it?

You're correct, the previous behavior was "standard" in the sense that I've seen many other applications work the same way -- double click loaded the track into the players, but click -pause- click will edit the field.  So this is "clicking on a selected row" which is distinct from a doubleclick.  (In Mixxx, doubleclicking would not trigger the edit behavior).  This standard behavior no longer works in mixxx because this PR removed it on purpose.

I can understand wanting to disable the click-to-edit behavior because it can be destructive in a live environment.  And I've used other applications where a key combination is required to edit a field, like ctrl-r when renaming a file for instance.  Sometimes the enter key is used to trigger editing.  There's no one solution.

shift-click is not standard and it could collide with multi-select, but it may be worth a try.  Or we could tweak the library preferences to add a checkbox for "clicking on selected row edits field" (distinct from double-click behavior).

Be-ing commented:
> shift-click is not standard and it could collide with multi-select

How about Alt + click?

daschuer commented:
> What do you mean by "selected click"? Clicking again on the track that is already focused? That would not work because edit mode would be triggered before double clicking on a track.

This is a standard behavior http://doc.qt.io/archives/qt-4.8/qabstractitemview.html#EditTrigger-enum and should work. In the Mixxx context it turns out as unwanted for some users https://bugs.launchpad.net/mixxx/+bug/1743238 I have seen the same discussion in the Nemo/Nautilus file manager.This is due to Linux allows to pen the file/folder with a single click.  

I think we should use an extra combo-box to allow it again like proposed here:
https://github.com/mixxxdj/mixxx/pull/1550#issuecomment-374419800

If this is not possible for 2.1, we should just revert this PR.

daschuer commented:
There is an other issue: The controller mapping "LoadSelectedIntoFirstStopped" will now start editing the track.

daschuer commented:
A possible follow up:
https://github.com/mixxxdj/mixxx/pull/1561

ywwg commented:
> There is an other issue: The controller mapping "LoadSelectedIntoFirstStopped" will now start editing the track.

ooh yeah that's definitely a problem
